File: Ejercicio_Clasificación/EjercicioClasificación.R
# Natural Language Processing

# Importar Dataset
dataset_original = read.delim("Restaurant_Reviews.tsv", quote = '', stringsAsFactors = FALSE)

# Limpieza de Texto
#install.packages('tm')
#install.packages('SnowballC')
library(tm)
library(SnowballC)
corpus = VCorpus(VectorSource(dataset_original$Review))
corpus = tm_map(corpus, content_transformer(tolower))
# Consultar el primer Elemento del Corpus
# as.character(corpus[[1]])
corpus = tm_map(corpus, removeNumbers)
corpus = tm_map(corpus, removePunctuation)
corpus = tm_map(corpus, removeWords, stopwords(kind = 'en')) 
corpus = tm_map(corpus, stemDocument)
corpus = tm_map(corpus, stripWhitespace)

# Crear el Bag of Words
dtm = DocumentTermMatrix(corpus)
dtm = removeSparseTerms(dtm, 0.999)

dataset = as.data.frame(as.matrix(dtm))
dataset$Liked = dataset_original$Liked

# Codificar a variable de clasificación como factor
dataset$Liked = factor(dataset$Liked,
                       levels = c(0,1))


# 01 Regresión Logística
# Dividir los datos en conjunto de entrenamiento y conjunto de testing
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Liked, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE )
testing_set = subset(dataset, split == FALSE)

# Ajustar el modelo de regresión logística con el conjunto de datos
classifier = glm(formula = Liked ~ .,
                 data = training_set, 
                 family = binomial)

# Predicción de los resultados con el conjunto de testing 
prob_pred = predict(classifier, type = 'response',
                    newdata = testing_set[,-692])
y_pred = ifelse(prob_pred>0.5, 1, 0)

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)


# 02 KNN
# Ajustar el clasificador con el conjunto de datos
# install.packages('class')
library(class)
y_pred = knn(train = training_set[,-692],
             test = testing_set[,-692],
             cl = training_set[,692],
             k = 5)

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)


# 03 SVM
# Ajustar el SVM con el conjunto de datos
# install.packages('e1071')
library(e1071)
classifier = svm(formula = Liked ~ .,
                 data = training_set,
                 type = 'C-classification',
                 kernel = 'linear')

# Predicción de los resultados con el conjunto de testing 
y_pred = predict(classifier, newdata = testing_set[,-692])

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)


# 04 Kernel SVM
# Ajustar el kernel SVM con el conjunto de datos
# install.packages('e1071')
library(e1071)
classifier = svm(formula = Liked ~ .,
                 data = training_set,
                 type = 'C-classification',
                 kernel = 'radial')

# Predicción de los resultados con el conjunto de testing 
y_pred = predict(classifier, newdata = testing_set[,-692])

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)


# 05 Naive Bayes
# Ajustar el Naive Bayes con el conjunto de datos
# install.packages('e1071')
library(e1071)
classifier = naiveBayes( x = training_set[,-692],
                         y = training_set$Liked)

# Predicción de los resultados con el conjunto de testing 
y_pred = predict(classifier, newdata = testing_set[,-692])

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)


# 06 Decission Tree
# Ajustar el clasificador de árbol de decisión con el conjunto de datos
# install.packages('rpart')
library(rpart)
classifier = rpart(formula = Liked ~ .,
                   data = training_set)

# Predicción de los resultados con el conjunto de testing 
y_pred = predict(classifier, newdata = testing_set[,-692],
                 type = 'class')

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)


# 07 Random Forest
# Ajustar el clasificador Random Forest con el conjunto de datos
# install.packages('randomForest')
library(randomForest)
classifier = randomForest(x = training_set[,-692],
                          y = training_set$Liked,
                          ntree = 10)

# Predicción de los resultados con el conjunto de testing 
y_pred = predict(classifier, newdata = testing_set[,-692])

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)


# 08 Cart
# Ajustar el clasificador de CART con el conjunto de datos
# install.packages('rpart')
library(rpart)
classifier = rpart(formula = Liked ~ .,
                   data = training_set,
                   method = 'class',
                   control = rpart.control(cp = 0.01))

# Predicción de los resultados con el conjunto de testing 
y_pred = predict(classifier, newdata = testing_set[,-692],
                 type = 'class')

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)


# 09 C5.0
# Ajustar el clasificador de C5.0 con el conjunto de datos
# install.packages('C50')
library(C50)
classifier = C5.0(formula = Liked ~ .,
                  data = training_set,
                  trials = 10)

# Predicción de los resultados con el conjunto de testing 
y_pred = predict(classifier, newdata = testing_set[,-692],
                 type = 'class')

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)


# 10 Max. Entropy
# Ajustar el clasificador de max ent con el conjunto de datos
# install.packages("nnet")
library(nnet)
classifier = multinom(formula = Liked ~ .,
                  data = training_set,
                  MaxNWts = 1000)

# Predicción de los resultados con el conjunto de testing 
y_pred = predict(classifier, newdata = testing_set[,-692],
                 type = 'class')

# Crear la matriz de confusión
cm = table(testing_set[,692], y_pred)

File: Ejercicio_Clasificación/EjercicioClasificación.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun 13 20:29:24 2023

@author: Ckonny
"""

# Ejercicio de Clasificación
# Importar librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar dataset
dataset = pd.read_csv("Restaurant_Reviews.tsv", delimiter="\t", quoting=3)

# Limpieza de Texto
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

corpus = []
for i in range(0, 1000):
    review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])
    review = review.lower()
    review = review. split()
    ps = PorterStemmer()
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review = ' '.join(review)
    corpus.append(review)
    
# Crear Bags of Words
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500)
X = cv.fit_transform(corpus).toarray()
y = dataset.iloc[:, 1].values


# 1 REGRESIÓN LOGÍSTICA
# Dividir el dataset en conjunto de entrenamiento y testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

# Ajustar el modelo de regresión logística al conjunto de entrenamiento
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train) 

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
                      

# 2 KNN
# Ajustar el clasificador modelo al conjunto de entrenamiento
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier.fit(X_train, y_train)

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)

# Elaborar matriz de confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)


# 3 SVM
# Ajustar el SVM modelo al conjunto de entrenamiento
from sklearn.svm import SVC
classifier = SVC(kernel = "linear", random_state = 0)
classifier.fit(X_train, y_train)

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de Confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)


# 4 KERNEL SVM
# Ajustar el kernel SVM modelo al conjunto de entrenamiento
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0)
classifier.fit(X_train, y_train)

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de Confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)


# 5 NAIVE BAYES
# Ajustar el Naive Bayes modelo al conjunto de entrenamiento
from sklearn.naive_bayes import GaussianNB 
classifier = GaussianNB()
classifier.fit(X_train, y_train)

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de Confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)


# 6 DECISSION TREE
# Ajustar el clasificador de árbol de decisión  modelo al conjunto de entrenamiento
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion='entropy', random_state=0)
classifier.fit(X_train, y_train)

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de Confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)


# 7 RANDOM FOREST
# Ajustar el clasificador de bosques aleatorios modelo al conjunto de entrenamiento
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de Confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)


# 8 CART 
# Ajustar el clasificador CART modelo al conjunto de entrenamiento
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion="entropy", max_depth=2)
classifier.fit(X_train, y_train)

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de Confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)


# 9 C5.0 
# Ajustar el clasificador C5.0 modelo al conjunto de entrenamiento
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion="entropy", min_samples_split=75, max_leaf_nodes=5 )
classifier.fit(X_train, y_train)

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de Confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)


# 10 MAX ENTROPY
# Ajustar el modelo de regresión logística (MAXEN) al conjunto de entrenamiento
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0, multi_class='multinomial', solver='saga')
classifier.fit(X_train, y_train) 

# Predicción con los resultados del conjunto de testing
y_pred = classifier.predict(X_test)
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)